pkg/sql/catalog/schematelemetry/schematelemetrycontroller/controller.go

```diff
@@ -0,0 +1,289 @@
+)
+
+// SchemaTelemetryScheduleName is the name of the schema telemetry schedule.
+const SchemaTelemetryScheduleName = "sql-schema-telemetry"
+
+// SchemaTelemetryRecurrence is the cron-tab string specifying the recurrence
+// for schema telemetry job.
+var SchemaTelemetryRecurrence = settings.RegisterValidatedStringSetting(
+	settings.TenantWritable,
+	"sql.schema.telemetry.recurrence",
+	"cron-tab recurrence for SQL schema telemetry job",
+	"@daily", /* defaultValue */
+	func(_ *settings.Values, s string) error {
+		if _, err := cron.ParseStandard(s); err != nil {
+			return errors.Wrap(err, "invalid cron expression")
+		}
+		return nil
+	},
+).WithPublic()
+
+// ErrDuplicatedSchedules indicates that there is already a schedule for SQL
+// schema telemetry jobs existing in the system.scheduled_jobs table.
+var ErrDuplicatedSchedules = errors.New("creating multiple schema telemetry schedules is disallowed")
+
+// ErrVersionGate indicates that SQL schema telemetry jobs or schedules are
+// not supported by the current cluster version.
+var ErrVersionGate = errors.New("SQL schema telemetry jobs or schedules not supported by current cluster version")
+
+// Controller implements the SQL Schema telemetry subsystem control plane.
+// This exposes administrative interfaces that can be consumed by other parts
+// of the database (e.g. status server, builtins) to control the behavior of the
+// SQL schema telemetry subsystem.
+type Controller struct {
+	scheduleController
+	jr *jobs.Registry
+}
+
+// scheduleController is distinct from Controller which implements the full
+// control plane, because tenant migrations don't have access to the jobs
+// registry.
+type scheduleController struct {
+	db *kv.DB
+	ie sqlutil.InternalExecutor
+	st *cluster.Settings
+}
+
+// NewController is a constructor for *Controller.
+//
+// This constructor needs to be called in the sql package when creating a new
+// sql.Server. This is the reason why it and the definition of the Controller
+// object live in their own package separate from schematelemetry.
+func NewController(
+	db *kv.DB, ie sqlutil.InternalExecutor, st *cluster.Settings, jr *jobs.Registry,
+) *Controller {
+	return &Controller{
+		scheduleController: scheduleController{
+			db: db,
+			ie: ie,
+			st: st,
+		},
+		jr: jr,
+	}
+}
+
+// CreateSchemaTelemetryJob implements eval.SchemaTelemetryController.
+func (c *Controller) CreateSchemaTelemetryJob(
+	ctx context.Context, createdByName string, createdByID int64,
+) (id int64, _ error) {
+	if !c.st.Version.IsActive(ctx, clusterversion.SQLSchemaTelemetryScheduledJobs) {
+		return 0, ErrVersionGate
+	}
+	if err := c.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		jobID, err := CreateSchemaTelemetryJob(ctx, c.jr, txn, createdByName, createdByID)
+		id = int64(jobID)
+		return err
+	}); err != nil {
+		return 0, err
+	}
+	c.jr.NotifyToAdoptJobs()
+	return id, nil
+}
+
+// CreateSchemaTelemetryJob creates a new schema telemetry job.
```

Comment:
It's worth noting that it creates it in an "adoptable" mode such that any node can adopt it. Using default job creation functions, the job will execute first on the local node.

Reply:
Honestly, it might be better to just have a function that creates the record and deal with job creation on a per-call-site basis. Then you don't have to have the discussion or pass as many arguments around.

Reply:
TIL, thanks. Good idea.